tough-kms/src/lib.rs

```diff
@@ -188,7 +196,130 @@ impl Sign for KmsRsaKey {
                 key_id: self.key_id.clone(),
             })?;
-        let signature = response.signature.context(error::SignatureNotFound)?;
-        Ok(signature.to_vec())
+        let signature = response
+            .signature
+            .context(error::SignatureNotFound)?
+            .to_vec();
+
+        // sometimes KMS produces a signature that is shorter than the modulus. in those cases,
+        // we have observed that openssl and KMS will both validate the signature, but ring will
+        // not. if we pad the beginning of the signature with zeros to make the signature exactly
+        // the same length as the modulus, then ring will verify the signature.
+        let signature = pad_signature(signature, self.modulus_size_bytes)?;
+        Ok(signature)
+    }
+}
+
+/// Parses the `CustomerMasterKeySpec` string returned by KMS, e.g. `RSA_3072` and returns the size
+/// of the modulus in bytes. For example `RSA_3072` has a modulus of 3072 bits, so the function will
+/// return 384 == (3072 / 8). If the parsed number is not divisible by 8, an error is returned.
+fn parse_modulus_length_bytes(spec: &str) -> error::Result<usize> {
+    // only RSA is currently supported
+    ensure!(
+        spec.starts_with("RSA_"),
+        error::BadCustomerMasterKeySpec { spec }
+    );
+    // prevent a panic if the string is precisely "RSA_"
+    ensure!(spec.len() > 4, error::BadCustomerMasterKeySpec { spec });
+    // extract the digits
+    let mod_len_str = &spec[4..];
+    // parse the digits
+    let mod_bits = mod_len_str
+        .parse::<usize>()
+        .context(error::BadCustomerMasterKeySpecInt { spec })?;
+    // make sure the modulus size is compatible with u8 bytes
+    ensure!(
+        mod_bits % 8 == 0,
+        error::UnsupportedModulusSize {
+            modulus_size_bits: mod_bits,
+            spec,
+        }
+    );
+    // convert to 8-bit bytes
+    Ok(mod_bits / 8)
+}
+
+/// * If the length of `signature` is less than `modulus_size_bytes`, this function will prepend the
+///   `signature` with zeros so that `signature.len() == modulus_size_bytes`.
+/// * If the `signature` already the same length as `modulus_size_bytes` then `signature` is
+///   returned unchanged.
+/// * If the `signature` is longer than `modulus_size_bytes`, an error is returned.
+fn pad_signature(signature: Vec<u8>, modulus_size_bytes: usize) -> error::Result<Vec<u8>> {
+    ensure!(
+        signature.len() <= modulus_size_bytes,
+        error::SignatureTooLong {
+            modulus_size_bytes,
+            signature_size_bytes: signature.len()
+        },
+    );
+    if signature.len() == modulus_size_bytes {
+        return Ok(signature);
     }
+    // we now know that the signature is shorter than the modulus
+    let padding_size: usize = modulus_size_bytes - signature.len();
+    let mut padded: Vec<u8> = vec![0; modulus_size_bytes];
+    padded.splice(padding_size..modulus_size_bytes, signature.iter().cloned());
+    Ok(padded)
```

Comment:
```suggestion
    signature.splice(..0, [0].repeat(padding_size));
    Ok(signature)
```
We have ownership of `signature`, so this seems simpler..?